storage/pruner/pruner.go

```diff
@@ -0,0 +1,374 @@
+package pruner
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/pebble"
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/storage"
+	pebbleStorage "github.com/onflow/flow-go/storage/pebble"
+)
+
+type PruneMethod int
+
+const (
+	// PruneByHeight prunes data using finalized block heights
+	// This is should be used when a db does not include a view index
+	PruneByHeight PruneMethod = iota + 1
+
+	// PruneByView prunes data using views
+	// This is the simpler and more efficient method, but requires the node to have a view index
+	PruneByView
```

Comment:
Instead of using the method for different strategy, I would suggest to use different engine core.

Both PruneByHeightCore and PruneByViewCore provide the same API (a single `Prune()` method) for the wrapper to call. The wrapper is a component manager. This way we don't have to mix two logic, and easy to write different set of tests for the cores. And the cores can have different dependencies if necessary.